test/instrumentation/modules/@aws-sdk/fixtures/use-s3-v3.js

```diff
@@ -0,0 +1,337 @@
+//    export AWS_ACCESS_KEY_ID=fake
+//    export AWS_SECRET_ACCESS_KEY=fake
+// See also: https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-envvars.html
+//
+// Usage:
+//    # Run against the default configured AWS profile, creating a new bucket
+//    # and deleting it afterwards.
+//    node use-s3-v3.js | ecslog
+//
+//    # Testing against localstack.
+//    docker run --rm -it -e SERVICES=s3 -p 4566:4566 localstack/localstack
+//    TEST_ENDPOINT=http://localhost:4566 node use-s3-v3.js | ecslog
+//
+//    # Use TEST_BUCKET_NAME to re-use an existing bucket (and not delete it).
+//    # For safety the bucket name must start with "elasticapmtest-bucket-".
+//    TEST_BUCKET_NAME=elasticapmtest-bucket-1 node use-s3.js | ecslog
+//
+// Output from a sample run is here:
+// https://gist.github.com/trentm/c402bcab8c0571f26d879ec0bcf5759c
+
+const apm = require('../../../../..').start({
+  serviceName: 'use-s3-v3',
+  captureExceptions: false,
+  centralConfig: false,
+  metricsInterval: 0,
+  cloudProvider: 'none',
+  stackTraceLimit: 4, // get it smaller for reviewing output
+  logLevel: 'info'
+})
+
+const crypto = require('crypto')
+const vasync = require('vasync')
+const assert = require('assert')
+const {
+  S3Client,
+  ListBucketsCommand,
+  CreateBucketCommand,
+  DeleteBucketCommand,
+  PutObjectCommand,
+  GetObjectCommand,
+  DeleteObjectCommand,
+  waitUntilBucketExists,
+  waitUntilObjectExists
+} = require('@aws-sdk/client-s3')
+
+const TEST_BUCKET_NAME_PREFIX = 'elasticapmtest-bucket-'
+
+// ---- support functions
+/**
+ * Gets the data or error from a promise and calls the callback function in a calssic node fashion
+ *
+ * @param {Promise<unknown>} promise the promise we want to capture result and error for the callback
+ * @param {*} cb node lile callback with the signature (err, data)
+ */
+function toCallback (promise, cb) {
+  promise.then(
+    function (data) { cb(null, data) },
+    function (err) { cb(err, null) }
+  )
+}
+
+// https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/clients/client-s3/index.html
+function useS3 (s3Client, bucketName, cb) {
+  const region = s3Client.config.region
+  const log = apm.logger.child({
+    'event.module': 'app',
+    endpoint: s3Client.config.endpoint,
+    bucketName,
+    region
+  })
+  const key = 'aDir/aFile.txt'
+  const content = 'hi there'
+
+  vasync.pipeline({
+    arg: {},
+    funcs: [
+      // https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/clients/client-s3/classes/listbucketscommand.html
+      // Limitation: this doesn't handle paging.
+      function listAllBuckets (arg, next) {
+        const command = new ListBucketsCommand({})
+
+        toCallback(s3Client.send(command), function (err, data) {
+          log.info({ err, data }, 'listBuckets')
+          // TODO: need to check why this assertion fails
+          // assert(apm.currentSpan === null,
+          //   'S3 span should NOT be a currentSpan in its callback')
```

Comment:
My first guess is that this will fixed if you change from using `agent.startSpan(...)` to `agent._instrumentation.createSpan(...)` -- as is now used in most (all?) instrumentations.

*Why* that makes a difference can be a long discussion, which we should have at some point. A starting point on using the Instrumentation API is the "Overview" section of https://github.com/elastic/apm-agent-nodejs/pull/2181

Reply:
Indeed it was. I'm going to have a look at that PR to understand better. Thanks!